### DISCUSSION TITLE
Simulation.Inject adds actor at the end of the interval

### DISCUSSION THREAD

**Discussion by michaelplavnik** [VIP-Logic/NBomber#747] [General]
I am trying to simulate scheduled task behavior, where one execution happens every 5 minutes, and run it for hours (30 min at least).

I attempted it as follows:

`Simulation.Inject(1, TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(30))`

Observation shows that actor is injected at the end of the interval (and not really 5 min interval, but somehow ~ 4 min 30 sec).

As body of the simulation might run more then 30 sec, it is confusing to interpret.

Any ideas?

**Comment by AntyaDev** [comment]
HI @michaelplavnik ,
That's a very good point. The explanation is simple: **injectionInterval** specifies the interval between injections but not a concrete time when injection will happen. It's done to spread the load granularry instead of send all requests in the same millisecond.
Why it's done like this? 
Imagine the user configured `Simulation.Inject(1_000, TimeSpan.FromSeconds(1), TimeSpan.FromMinutes(30))` 
In this case, if to follow the flow that you described, NBomber will need to fire 1000 requests at the end of each second, which means we will send a batch of requests at the exact same time. The problem with such an approach is that the workload will produce a lot of spikes on the target system. NBomber introduces some random delay (during injection interval) for each request to spread the load and make it more granular/realistic.
```csharp
var startDelay = _randomizer.Next(0, int injectInterval.TotalMilliseconds)
Run(scenario, startDelay);
```

I think we can introduce some settings for NBomber to schedule requests at the exact time.



I think we probably need to explain this nuance in documentation.

**Comment by AntyaDev** [comment]
One more comment: 
`Simulation.Inject(1, TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(30))`
NBomber will split the duration (30 min) into intervals of 5 min each. 
In one of each interval, it will fire exactly one request. 
When the interval chunk starts, NBomber will schedule a request with random delay. The above I explained why random delay is important. Basically, when multiple requests are fired at the same moment, it will introduce spikes. Random delay helps to spread the load during interval chunk.

**Reply by michaelplavnik** [reply]
It would be great if one can configure function (int actorIdx, randomizer) that returns delay

**Reply by AntyaDev** [reply]
I am not sure it's the best option to expose such low-level functionality; even knowing about actors seems like a super low-level abstraction.
What do you think about providing some configuration to enable or disable this behavior? Seems quite simple for the user.

**Reply by AntyaDev** [reply]
I also can add that we have a task in our backlog to provide very custom scheduling, which you will be able to control, introduce any delays, etc. So basically you will be able to build your custom LoadSimulation.

**Reply by michaelplavnik** [reply]
Together with custom simulator option, simple flag is sufficient.

**Comment by michaelplavnik** [comment]
Meanwhile, could you please suggest how can I simulate scheduled task as described?

**Reply by AntyaDev** [reply]
It's currently impossible to disable randomness for scheduling. On the other hand, we can implement the configuration for disabling randomness, which doesn't seem a big deal at all.

**Reply by michaelplavnik** [reply]
I tried to approximate with
for (int i  = 0; i < invocations; ++i)
{
    simulations[i] = Simulation.Inject(1, TimeSpan.FromSeconds(10), TimeSpan.FromMinutes(5));
}
But is does not solve the issue, cause as soon as scenario completes, simulation completes.

**Reply by AntyaDev** [reply]
@michaelplavnik, as I mentioned, this randomness can't be disabled in the current NBomber version. I think we can add this feature in the next release if this use case is super important for your company.

**Reply by AntyaDev** [reply]
@michaelplavnik, can you please submit a talk about this feature?
